src/providers/data_provider/dp_target_id.c

```diff
@@ -448,10 +448,238 @@ static void dp_req_initgr_pp_set_initgr_timestamp(struct dp_initgr_ctx *ctx,
     }
 }
 
+
+struct dp_sr_resolve_groups_state {
+    struct data_provider *provider;
+    struct dp_initgr_ctx *initgroups_ctx;
+    struct dp_reply_std reply;
+
+    uint32_t *resolve_gids; /* Groups needing resolution */
+    int resolve_gnum;
+    int num_iter;
+    uint32_t gnum;
+};
+
+static errno_t dp_sr_resolve_groups_check(struct dp_sr_resolve_groups_state *state);
+static errno_t dp_sr_resolve_groups_next(struct tevent_req *req);
+static void dp_sr_resolve_groups_done(struct tevent_req *subreq);
+
+struct tevent_req *
+dp_sr_resolve_groups_send(TALLOC_CTX *mem_ctx,
+                          struct tevent_context *ev,
+                          struct dp_reply_std reply,
+                          struct data_provider *provider,
+                          struct dp_initgr_ctx *initgr_ctx)
+{
+
+    struct dp_sr_resolve_groups_state *state;
+    struct tevent_req *req;
+    int ret;
+    struct session_recording_conf sr_conf;
+
+    req = tevent_req_create(mem_ctx, &state, struct dp_sr_resolve_groups_state);
+    if (req == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create tevent request!\n");
+        return NULL;
+    }
+
+    if (initgr_ctx->username == NULL) {
+        ret = EOK;
+        goto done;
+    }
+
+    sr_conf = provider->be_ctx->sr_conf;
+
+    /* Only proceed if scope is applicable: 'some' or 'all' with groups to resolve */
+    if ((sr_conf.scope == SESSION_RECORDING_SCOPE_SOME && sr_conf.groups != NULL)
```

Comment:
It would be nicer to avoid the else branch, e.g.

```
    if (condition_to_return_eok) {
        ret = EOK;
        goto done;
    }

    do stuff

    return req;

done:
...
```